### DISCUSSION TITLE
Feature request: SWC Plugins in Next.js

### DISCUSSION THREAD

**Discussion by tomitrescak** [vercel/next.js#30212] [Ideas]
### Describe the feature you'd like to request

I would like to use SWC plugins

### Describe the solution you'd like

I am old-school and I'm using jsx-components (If, Else). With babel there is a pretty great plugin for it (https://www.npmjs.com/package/babel-plugin-jsx-control-statements). 

I wrote one for SWC and I would love to use it with Next.js

### Describe alternatives you've considered

None

**Comment by showmeyourhits** [comment]
Is there a way to customize swc configuration with your own plugins?

We have our own babel plugins and could refactor them to swc, but I coudn't find any info on how to use them - neither in nextjs docs nor in [swc](https://swc.rs/docs/usage/plugins)

**Comment by jlopezxs** [comment]
Same here.
I'm using a custom babel plugin to create data attributes (data-qa) with the displayName value in kebabCase, but migrating to next12 with swc I have not found the way to attach other plugins in the transforms.

Is in the roadmap provide this API to add plugins in next.config? or right now there is something that I have not seen yet.

**Reply by philwolstenholme** [reply]
Did you ever get something like this ("create data attributes (data-qa) with the displayName value in kebabCase") working with SWC @jlopezxs ? I'm looking for something similar

**Comment by LukasBombach** [comment]
I keep hitting walls with a bunch of side- and work-related projects that require or benefit from compile-time code-transformations. These things don‘t seem to be reasonably doable with next 12. Since next is a major player in our eco-system I assume a lot of library authors face the same dilemma.

@leerob is there any roadmap or discussion you can share on enabling this matter?

**Comment by leerob** [comment]
We're still doing research on a solution for plugins with SWC and Next.js 👍

**Reply by LukasBombach** [reply]
coolz thanks

**Reply by jantimon** [reply]
@leerob could you please give an update on this topic? 

In the meantime a new discussion has been started around the new SWC WASM plugins:
https://github.com/vercel/next.js/discussions/34384

**Reply by mdarche** [reply]
@leerob curious to hear more about this as well

**Reply by jantimon** [reply]
vercel merged https://github.com/vercel/next.js/pull/36790 and released it as npm canary version [12.1.7-canary.27](https://www.npmjs.com/package/next/v/12.1.7-canary.27)

here is an example how you can now use **experimental** swc wasm plugins:  
https://github.com/jantimon/next-swc-plugin-demo/blob/9489ab5902c5a1032ddde0e178f2daac24b22e5b/next.config.js#L4-L7

beside some minor bugs (https://github.com/vercel/next.js/issues/37409) it works very well so far 👍

**Reply by mdarche** [reply]
Exactly what I was looking for-- thanks for the intel @jantimon!

**Comment by chrisrichard** [comment]
@tomitrescak did you get JSX Control Statements working in Next.js/SWC? This is the only thing preventing us from using the SWC compiler...

**Reply by tomitrescak** [reply]
Hello, we actually removed most of the tech that required any kind of setup to reduce the upgrade friction. We have only base config for some dev tool such as prettier, but all the code generation has 0 config apart from tsconfig. To do that we removed all the code and developed some helper components such as the one for For. So no ... instead of bending it our way .. we bend their way ;)
